### ISSUE TITLE
Injecting a parent’s custom attribute doesn’t work

### ISSUE THREAD

**Issue by rluba** [aurelia/aurelia#660]
# 🐛 Bug Report

@fkleuver, as discussed in Discord: I converted two custom attributes from Aurelia 1 to Aurelia 2. One must be nested inside the other.

Usage looks like this:
```HTML
<li au-dropdown>
	<a au-dropdown-toggle>…</a>
	…
</li>
```
The inner attribute in V1 (works):
```JavaScript
@inject(DOM.Element, AuDropdownCustomAttribute)
export class AuDropdownToggleCustomAttribute {
```

When I try to inject the outer attribute into the inner one, Aurelia v2 injects a _new_ instance of the outer attribute (and doesn’t initialize the bindings at all):
```TypeScript
export class AuDropdownToggleCustomAttribute {
    constructor(public element: HTMLElement, public dropdown: AuDropdownCustomAttribute) {
```

## 🤔 Expected Behavior

The constructor should get the attribute instance from its parent element, as it did in Aurelia 1.

**Comment by EisenbergEffect** [comment]
I think this is an important use case, but I'm not sure we should enable it in the same was that we did in v1. Here are a few quick thoughts. @rluba I'd love to hear what you think. FYI for @fkleuver as well.

In v1, we created a dedicated container instance for every DOM node instance that was a custom element or had a custom attribute on it. This meant there was a 1:1 mapping between container instances and Aurelia behaviors. Anything created out of one of these containers was a singleton. This enabled the pattern above, but at the cost of creating many, many DI containers, which all needed to be kept in memory. So, there were both perf and memory concerns there. It also made me feel pretty uncomfortable, since the one unique scenario it seemed to enable was the one above. It's a lot of cost on the entire system for just this one thing.

For v2 we're doing some different things with containers that enable us to have fewer of them. I think that's a good thing if we can keep it that way. So, how to address this use case?

Well, we've got a `@children` decorator. Why not add a `@parent` decorator? Since this is pretty DOM-centric, does it really need to be done with DI? My thought is that this is better handled through our component system and connected with lifecycle hooks.

Regarding lifecycle, here's something that may not be immediately apparent. The v1 implementation, based on DI, has a pretty big flaw. If you move the child element around in the DOM, say with a drag drop repeater or really any other sort of template control that removes something, you are stuff with the previously injected parent attribute, which is no longer valid based on the new DOM position. If we implement a lifecycle based solution, something similar to @children` or provide some sort of API specifically for this, then we can disconnect and reconnect parent behaviors when things are moved.

So, in summary: We want to support this use case, but I think this may be a time when we want to make a breaking change and enable this through a more lifecycle-integrated mechanism.

**Comment by rluba** [comment]
> If you move the child element around in the DOM…
I haven’t had that problem, ever. It’s always just been "give me a parent/child during setup time". I don’t know if this is something others have needed in the past.

Regardless, I think your approach sounds very sensible. I really don’t care if I get my parent/children via DI or via the component system. Whatever’s easier and/or faster.

P.S.: `@parent` seems more straight-forward to me than `@children`, but I can work with either one. So there’s no need to implement `@parent` if the other one already exists and works.

**Comment by EisenbergEffect** [comment]
FYI `@children` is already implemented for elements that want to get a dynamic list of child components. For example a tab controls that wants an always synced array of child tab components. So, in this case, we'd be enabling the inverse wireup.

I'm sure @fkleuver will have some technical thoughts on all of this as well. He may tell me I'm completely wrong anyway and my concerns aren't relevant anymore 🤣 I'm so glad to have such an excellent team these days. When vNext is ready, it's going to be awesome.

**Comment by fkleuver** [comment]
We ended up with more containers than we started with in v2, a bit of a middle ground it seems. Not quite "container per component", but rather "container per component type per hierarchy". Remember this pr: https://github.com/aurelia/aurelia/pull/627/files#diff-b12a454f7d84dd07c94592e2fb3b0a30R269

Say you've got this component tree:
```
- Foo
  - Bar 1
  - Bar 2
- Baz
  - Bar 3
  - Bar 4
```

In v1 this would be 6 containers. Early v2 this would have been 3 containers: 1 per component type.
Now we will have 4 containers. One for `Foo`, one for `Bar 1` and `Bar 2` (same type under same parent), one for `Baz`, and one for `Bar 3` and `Bar 4` (same type under different parent).

With this structure we can, to the best of my knowledge, safely add the parent instance registrations and enable this scenario.

That said, you can also grab the parent by walking up the controller tree:

```ts
this.$controller.parent.viewModel // the immediate parent component
```

**Comment by rluba** [comment]
`@children` doesn’t work in my case because both the parent and the children are _attributes_, not elements. And apparently, `@children` requires an element at the moment – which I didn’t expect.

**Comment by EisenbergEffect** [comment]
I definitely think we can and should fix that. The underlying functionality only relies on a projector being present. Only elements have those but we can implement a simple one for attributes that is available on-demand for this scenario. The project is the abstraction through which the children decorator finds out what elements are the children, more or less, since different element projector types might handle this differently (and an attribute would potentially as well).

**Comment by Sayan751** [comment]
@fkleuver I think this can be easily solvable by making the `link` a formal lifecycle hook for custom attributes. That gives devs enormous flexibility to link whatever parent controller or view model for that matter. Although the API might not be very nicer, I think that it might be efficient from the framework standpoint.

Ofc having `@parent` decorator would also be nice, to property-inject the instances around the same time `link` happens. I think the metadata for that can be used quite efficiently.

cc: @bigopon

**Comment by bigopon** [comment]
This is now supported via `link` lifecycle in custom attributes (normal & template controller).

**Comment by bigopon** [comment]
There's a relevant discourse at https://discourse.aurelia.io/t/au2-access-parent-component-viewmodel-in-a-clean-way/5378/5 for those who come later here.
This part probably will go into the FAQ, with more ways to do it via container.

**Comment by bigopon** [comment]
It felt like there's still a lack of capability with `link`, so I've created a new PR for walking + finding custom attributes, the PR is at #1928 . If anyone interested, please have a look there.
